File: gospider/research/notebook_output.go
package research

import (
	"fmt"
	"time"
)

type ExperimentRecord struct {
	ID        string                 `json:"id"`
	Name      string                 `json:"name"`
	Timestamp float64                `json:"timestamp"`
	URLs      []string               `json:"urls"`
	Schema    map[string]interface{} `json:"schema"`
	Results   []map[string]interface{} `json:"results"`
	Metadata  map[string]interface{} `json:"metadata"`
}

type ExperimentTracker struct {
	Experiments []ExperimentRecord `json:"experiments"`
}

func NewExperimentTracker() *ExperimentTracker {
	return &ExperimentTracker{Experiments: []ExperimentRecord{}}
}

func (t *ExperimentTracker) Record(
	name string,
	urls []string,
	results []map[string]interface{},
	schema map[string]interface{},
	metadata map[string]interface{},
) ExperimentRecord {
	record := ExperimentRecord{
		ID:        fmt.Sprintf("exp-%03d", len(t.Experiments)+1),
		Name:      name,
		Timestamp: float64(time.Now().Unix()),
		URLs:      append([]string{}, urls...),
		Schema:    cloneMap(schema),
		Results:   cloneRows(results),
		Metadata:  cloneMap(metadata),
	}
	t.Experiments = append(t.Experiments, record)
	return record
}

func (t *ExperimentTracker) GetExperiment(name string) *ExperimentRecord {
	for index := range t.Experiments {
		if t.Experiments[index].Name == name {
			return &t.Experiments[index]
		}
	}
	return nil
}

func (t *ExperimentTracker) Compare() map[string]interface{} {
	experiments := make([]map[string]interface{}, 0, len(t.Experiments))
	totalURLs := 0
	totalResults := 0
	for _, experiment := range t.Experiments {
		totalURLs += len(experiment.URLs)
		totalResults += len(experiment.Results)
		experiments = append(experiments, map[string]interface{}{
			"id":               experiment.ID,
			"name":             experiment.Name,
			"urls_count":       len(experiment.URLs),
			"results_count":    len(experiment.Results),
			"success_rate":     calculateSuccessRate(experiment.Results),
			"avg_extract_time": averageExtractTime(experiment.Results),
		})
	}
	return map[string]interface{}{
		"experiments": experiments,
		"summary": map[string]interface{}{
			"total_experiments": len(t.Experiments),
			"total_urls":        totalURLs,
			"total_results":     totalResults,
		},
	}
}

func (t *ExperimentTracker) ToRows() []map[string]interface{} {
	rows := []map[string]interface{}{}
	for _, experiment := range t.Experiments {
		for _, result := range experiment.Results {
			rows = append(rows, map[string]interface{}{
				"experiment_id":   experiment.ID,
				"experiment_name": experiment.Name,
				"seed":            result["seed"],
				"extract":         result["extract"],
				"duration_ms":     result["duration_ms"],
				"error":           result["error"],
			})
		}
	}
	return rows
}

func calculateSuccessRate(results []map[string]interface{}) float64 {
	if len(results) == 0 {
		return 0
	}
	success := 0
	for _, result := range results {
		if text, ok := result["error"].(string); !ok || text == "" {
			success++
		}
	}
	return float64(success) / float64(len(results)) * 100
}

func averageExtractTime(results []map[string]interface{}) float64 {
	if len(results) == 0 {
		return 0
	}
	var total float64
	var count float64
	for _, result := range results {
		switch value := result["duration_ms"].(type) {
		case float64:
			total += value
			count++
		case int:
			total += float64(value)
			count++
		}
	}
	if count == 0 {
		return 0
	}
	return total / count
}

func cloneMap(source map[string]interface{}) map[string]interface{} {
	if source == nil {
		return map[string]interface{}{}
	}
	clone := make(map[string]interface{}, len(source))
	for key, value := range source {
		clone[key] = value
	}
	return clone
}

func cloneRows(source []map[string]interface{}) []map[string]interface{} {
	rows := make([]map[string]interface{}, 0, len(source))
	for _, row := range source {
		rows = append(rows, cloneMap(row))
	}
	return rows
}

File: rustspider/src/notebook_output.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExperimentRecord {
    pub id: String,
    pub name: String,
    pub timestamp: f64,
    pub urls: Vec<String>,
    pub schema: Value,
    pub results: Vec<Value>,
    pub metadata: Value,
}

#[derive(Debug, Default)]
pub struct ExperimentTracker {
    pub experiments: Vec<ExperimentRecord>,
}

impl ExperimentTracker {
    pub fn record(
        &mut self,
        name: &str,
        urls: Vec<String>,
        results: Vec<Value>,
        schema: Option<Value>,
        metadata: Option<Value>,
    ) -> ExperimentRecord {
        let record = ExperimentRecord {
            id: format!("exp-{:03}", self.experiments.len() + 1),
            name: name.to_string(),
            timestamp: std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .map(|value| value.as_secs_f64())
                .unwrap_or_default(),
            urls,
            schema: schema.unwrap_or_else(|| serde_json::json!({})),
            results,
            metadata: metadata.unwrap_or_else(|| serde_json::json!({})),
        };
        self.experiments.push(record.clone());
        record
    }

    pub fn get_experiment(&self, name: &str) -> Option<&ExperimentRecord> {
        self.experiments.iter().find(|item| item.name == name)
    }

    pub fn compare(&self) -> Value {
        let experiments = self
            .experiments
            .iter()
            .map(|experiment| {
                let success_count = experiment
                    .results
                    .iter()
                    .filter(|result| result.get("error").is_none() || result["error"].as_str() == Some(""))
                    .count();
                let average_duration = experiment
                    .results
                    .iter()
                    .filter_map(|result| result.get("duration_ms").and_then(|value| value.as_f64()))
                    .collect::<Vec<_>>();
                serde_json::json!({
                    "id": experiment.id,
                    "name": experiment.name,
                    "urls_count": experiment.urls.len(),
                    "results_count": experiment.results.len(),
                    "success_rate": if experiment.results.is_empty() { 0.0 } else { success_count as f64 * 100.0 / experiment.results.len() as f64 },
                    "avg_extract_time": if average_duration.is_empty() { 0.0 } else { average_duration.iter().sum::<f64>() / average_duration.len() as f64 },
                })
            })
            .collect::<Vec<_>>();
        serde_json::json!({
            "experiments": experiments,
            "summary": {
                "total_experiments": self.experiments.len(),
                "total_urls": self.experiments.iter().map(|item| item.urls.len()).sum::<usize>(),
                "total_results": self.experiments.iter().map(|item| item.results.len()).sum::<usize>(),
            }
        })
    }

    pub fn to_rows(&self) -> Vec<Value> {
        let mut rows = Vec::new();
        for experiment in &self.experiments {
            for result in &experiment.results {
                rows.push(serde_json::json!({
                    "experiment_id": experiment.id,
                    "experiment_name": experiment.name,
                    "seed": result.get("seed").cloned().unwrap_or(Value::Null),
                    "extract": result.get("extract").cloned().unwrap_or_else(|| serde_json::json!({})),
                    "duration_ms": result.get("duration_ms").cloned().unwrap_or(Value::Null),
                    "error": result.get("error").cloned().unwrap_or_else(|| Value::String(String::new())),
                }));
            }
        }
        rows
    }
}
